### ISSUE TITLE
Failed to parse time string with nanoseconds in OpenStack\Common\Resource\Alias

### ISSUE THREAD

**Issue by zarv1k** [php-opencloud/openstack#215]
Hi!
I'm trying to implement authentication into Selectel Cloud Storage according to v3 example described in [Selectel Knowledge Base](https://kb.selectel.ru/22058988.html#APIоблачногохранилища-Аутентификация). There is working curl example on this page:

```
$ curl -i  https://api.selcdn.ru/v3/auth/tokens -XPOST -d '{"auth": { "identity": { "methods": ["password"], "password": { "user": { "id": "43371", "password": "QyjaoXLh"}}}}}'
HTTP/1.1 201 Created
Access-Control-Allow-Origin: *
Content-Length: 811
Content-Type: application/json
X-Subject-Token: 79aa8d7809948265c66d853cdd9bcc36
Date: Fri, 16 Feb 2018 14:31:47 GMT

{"token":{"expires_at":"2018-02-17T17:31:47.327877340Z","issued_at":"2018-02-16T17:31:47.327877340Z","methods":["password"],"project":{"domain":{}},"Catalog":[{"endpoints":[{"id":"79aa8d7809948265c66d853cdd9bcc36","region_id":"RegionOne","url":"https://api.selcdn.ru/v1/SEL_43371","region":"RegionOne","interface":"public"},{"id":"79aa8d7809948265c66d853cdd9bcc36","region_id":"RegionOne","url":"https://api.selcdn.ru/v1/SEL_43371","region":"RegionOne","interface":"admin"},{"id":"79aa8d7809948265c66d853cdd9bcc36","region_id":"RegionOne","url":"https://api.selcdn.ru/v1/SEL_43371","region":"RegionOne","interface":"internal"}],"type":"object-store","name":"swift","id":""}],"user":{"id":"79aa8d7809948265c66d853cdd9bcc36","name":"43371","domain":{"id":"default","name":"Default","links":{}}},"audit_ids":[""]}}
```

When I tried to implement authentication in my app using this library, I got the following error:
`Exception: DateTimeImmutable::__construct(): Failed to parse time string (2018-02-17T16:04:33.538862758Z) at position 0 (2): The timezone could not be found in the database in /app/vendor/php-opencloud/openstack/src/Common/Resource/Alias.php:54`

As you can see, Selectel Cloud uses dates in date format with nanoseconds, which causes a problem with creating DateTimeImmutable objects.

Can I override Alias class implementation somehow for fix that problem or could you please suggest any solution for that?

Thank you.

```
$ php -v
PHP 7.1.7 (cli) (built: Jul 15 2017 18:08:09) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
```

**Comment by haphan** [comment]
Indeed this is a limitation of the current PHP not allowing decimal fraction of a second.

Drupal project handles this by coming up with its own [ISO8601 parser](https://api.drupal.org/api/drupal/vendor%21zendframework%21zend-stdlib%21src%21DateTime.php/8.2.x).

I don't see any easy fix for this but to do exactly the same.

Btw, what is the version of Openstack is this? @zarv1k

**Comment by haphan** [comment]
Apparently PHP can only handle up to microsecond resolution. 

Any solution to this will result in data loss.

**Comment by zarv1k** [comment]
@haphan its fresh version I installed via composer a few hours before - v3.0.3.

**Comment by zarv1k** [comment]
I figured out that there is no good solution for this that could be used in this library for parsing dates without data losing. Any broken solution shouldn't be used in core of any library. But I think it would be great if the library supported overriding of date creation implementation somehow, until dates in such formats would be supported in future versions of PHP natively.
Anyway, I'm almost sure that in my case It does not matter if I'll cut precision in dirty way till the supported one by PHP parser. I see the only one solution just for myself - to use own fork where I could use crutch solution just for my concrete case only.

**Comment by haphan** [comment]
@zarv1k  I meant the version of OpenStack itself (Micata, Newton, Ocata, etc..).

I think this is a valid issue and need to be addressed by this project. 

Checkout #216 and let me know your opinion.

**Comment by zarv1k** [comment]
@haphan actually I don't know what version of OpenStack is used by Selectel Cloud Storage service. I didn't find any info about that. I don't know how can I help you with that. It's just a service I need to integrate with in my app.

According to changes in#216 - looks good. I've already tried to apply them and I don't see any issues with parsing dates at the moment.
But I started experiencing issues after dates parsing stage. 
`'Error: Call to a member function getServiceUrl() on null in /app/vendor/php-opencloud/openstack/src/Identity/v3/Service.php:49`
Could you please look at the example curl response json provided in issue description. There is 'Catalog' key in response with first letter in upper case that does not fit the corresponding alias for catalog in `/app/vendor/php-opencloud/openstack/src/Identity/v3/Models/Token.php:58` and leads to error above. I started thinking that this cloud service does not follow API v3 specification but I'm not sure.

**Comment by haphan** [comment]
Yep, looks like this is a non-standard response. 

You can try to fork this repo and modify the code to accept `Catalog` key.

**Comment by haphan** [comment]
Let me ping @tnt-dev if he can help.

@tnt-dev: We got a few discrepancies in  the JSON response from `https://api.selcdn.ru/v3/auth/tokens`
Can you share with us what version of OpenStack Selectel is using? What is the back-end of swift object storage service? THanks

**Comment by haphan** [comment]
Looks like it is a go-lang adaptation for swift api [Hummingbird](https://blog.selectel.com/cloud-storage-api-updates/#more-5456). 
It would be nice if it can follow the standard though.

**Comment by zarv1k** [comment]
I tried to change 'Catalog' with 'catalog' and got other errors. After unsuccessful playing with code changes, I just switched to from v3 to v2 auth endpoint and used sample for identity v2 and it works well out of box. In v2 there are no dates with precision even in milliseconds. When I authenticated using v2 endpoint, I tried using the other API methods using this library like object list, get object, object metadata, etc. and didn't find any problems with them.
So /v3/auth/tokens responses from Selectel in fact are non-standard.
